### ISSUE TITLE
xfreerdp: double free or corruption

### ISSUE THREAD

**Issue by execlogic** [FreeRDP/FreeRDP#2439]
Hi,
I just try the latest xfreerdp client version git 1.1.0-beta+2013071101-159-g27545 
Client (kubuntu 14.04 64-bit, 3.13.0-45-generic) connect to server : 
xfreerdp +clipboard -wallpaper:off -themes:off -encryption:off /kbd-type:pc105 /u:$USER /size:1200x800 /cert-ignore /v:192.168.20.100
After work few minutes client is crushed.

gdb:
[New Thread 0x7fffebfff700 (LWP 14032)]
*** Error in `/usr/bin/xfreerdp': malloc(): smallbin double linked list corrupted: 0x00007fffec052e70 ***

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff0c56700 (LWP 14031)]
0x00007ffff7183cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56      ../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.

(gdb) bt
#0  0x00007ffff7183cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff71870d8 in __GI_abort () at abort.c:89
#2  0x00007ffff71c0f24 in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff72cf6c8 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff71cbc87 in malloc_printerr (action=<optimized out>, str=0x7ffff72cfa90 "malloc(): smallbin double linked list corrupted", ptr=<optimized out>) at malloc.c:4996
#4  0x00007ffff71ce994 in _int_malloc (av=0x7fffec000020, bytes=16) at malloc.c:3359
#5  0x00007ffff71d0340 in __GI___libc_malloc (bytes=16) at malloc.c:2891
#6  0x00007ffff3b77e42 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#7  0x00007ffff3b78198 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#8  0x00007ffff6e53bb8 in ?? () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#9  0x00007ffff6e53cfc in ?? () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#10 0x00007ffff6e53fcd in _XEventsQueued () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#11 0x00007ffff6e4612d in XPending () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#12 0x00007ffff7b52dfe in xf_process_x_events (instance=instance@entry=0x6041d0) at /opt/repo/freerdp-1.1.0~git20140921.1.440916e+dfsg1/client/X11/xf_client.c:408
#13 0x00007ffff7b53c18 in xf_thread (param=0x6041d0) at /opt/repo/freerdp-1.1.0~git20140921.1.440916e+dfsg1/client/X11/xf_client.c:1416
#14 0x00007ffff49b0182 in start_thread (arg=0x7ffff0c56700) at pthread_create.c:312
#15 0x00007ffff724800d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

**Comment by execlogic** [comment]
Valgrind

==12578== Memcheck, a memory error detector
==12578== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==12578== Using Valgrind-3.10.0.SVN and LibVEX; rerun with -h for copyright info
==12578== Command: xfreerdp +clipboard -wallpaper:off -themes:off -encryption:off /kbd-type:pc105 /u:savko /size:1200x800  /cert-ignore /v:192.168.20.100
==12578== 
==12578== Warning: ignored attempt to set SIGKILL handler in sigaction();
==12578==          the SIGKILL signal is uncatchable
==12578== Warning: ignored attempt to set SIGSTOP handler in sigaction();
==12578==          the SIGSTOP signal is uncatchable
loading channel cliprdr
connected to 192.168.82.200:3389
connected to 192.168.82.200:3389
connected to 192.168.82.200:3389
Unable to get current timezone rule
==12578== Thread 2:
==12578== Invalid free() / delete / delete[] / realloc()
==12578==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x74F4F91: glyph_cache_put (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x74F5084: update_gdi_cache_glyph (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x6A6853D: update_recv_secondary_order (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A7575D: fastpath_recv_update.isra.11 (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A75B4B: fastpath_recv_updates (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A749CD: rdp_recv_pdu (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A77D27: transport_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68DC6: freerdp_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4CF33: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578==  Address 0xe549e00 is 0 bytes inside a block of size 28 free'd
==12578==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x74F4F91: glyph_cache_put (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x74F57B0: update_gdi_fast_glyph (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x6A68027: update_recv_primary_order (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A7575D: fastpath_recv_update.isra.11 (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A75B4B: fastpath_recv_updates (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A749CD: rdp_recv_pdu (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A77D27: transport_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68DC6: freerdp_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4CF33: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
Closed from X11
Server does not support XI 2.2
==12578== 
==12578== HEAP SUMMARY:
==12578==     in use at exit: 49,117 bytes in 587 blocks
==12578==   total heap usage: 960,789 allocs, 960,208 frees, 614,113,614 bytes allocated
==12578== 
==12578== Thread 1:
==12578== 13 bytes in 1 blocks are definitely lost in loss record 3 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x57873C9: strdup (strdup.c:42)
==12578==    by 0x826A73D: _strdup (in /usr/lib/x86_64-linux-gnu/libwinpr-crt.so.0.1.0)
==12578==    by 0x6A75009: tcp_get_ip_address (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A75267: tcp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A773BF: transport_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A555BF: nego_tcp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A555D8: nego_transport_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A55F7F: nego_attempt_nla (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56144: nego_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A71C42: rdp_client_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68A87: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578== 
==12578== 13 bytes in 1 blocks are definitely lost in loss record 4 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x57873C9: strdup (strdup.c:42)
==12578==    by 0x826A73D: _strdup (in /usr/lib/x86_64-linux-gnu/libwinpr-crt.so.0.1.0)
==12578==    by 0x6A75009: tcp_get_ip_address (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A75267: tcp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A773BF: transport_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A555BF: nego_tcp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A555D8: nego_transport_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A55FEF: nego_attempt_tls (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56144: nego_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A71C42: rdp_client_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68A87: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578== 
==12578== 20 bytes in 1 blocks are definitely lost in loss record 5 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x74F5433: update_process_glyph_fragments (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x74F5651: update_gdi_fast_index (in /usr/lib/x86_64-linux-gnu/libfreerdp-cache.so.1.1.0)
==12578==    by 0x6A68027: update_recv_primary_order (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A7575D: fastpath_recv_update.isra.11 (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A75B4B: fastpath_recv_updates (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A749CD: rdp_recv_pdu (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A77D27: transport_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68DC6: freerdp_check_fds (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4CF33: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== 48 bytes in 1 blocks are definitely lost in loss record 20 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x52F7760: CreateMutexW (in /usr/lib/x86_64-linux-gnu/libwinpr-synch.so.0.1.0)
==12578==    by 0x4E4A82A: xf_pre_connect (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x6A68A01: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4C9D5: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== 48 (16 direct, 32 indirect) bytes in 1 blocks are definitely lost in loss record 21 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x5AEF7F9: XGetModifierMapping (in /usr/lib/x86_64-linux-gnu/libX11.so.6.3.0)
==12578==    by 0x4E48BFD: xf_kbd_init (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x4E4AB63: xf_pre_connect (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x6A68A01: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4C9D5: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== 72 bytes in 1 blocks are definitely lost in loss record 54 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x50DC254: freerdp_client_context_new (in /usr/lib/x86_64-linux-gnu/libfreerdp-client.so.1.1.0)
==12578==    by 0x401682: main (in /usr/bin/xfreerdp)
==12578== 
==12578== 104 bytes in 1 blocks are definitely lost in loss record 60 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x54FC306: CreateThread (in /usr/lib/x86_64-linux-gnu/libwinpr-thread.so.0.1.0)
==12578==    by 0x4E4B29D: xfreerdp_client_start (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x4016DE: main (in /usr/bin/xfreerdp)
==12578== 
==12578== 1,072 (48 direct, 1,024 indirect) bytes in 1 blocks are definitely lost in loss record 78 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x7920E33: Stream_New (in /usr/lib/x86_64-linux-gnu/libwinpr-utils.so.0.1.0)
==12578==    by 0x6A55E84: nego_recv_response (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A55FA3: nego_attempt_nla (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56144: nego_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A71C42: rdp_client_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68A87: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4C9D5: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== 1,072 (48 direct, 1,024 indirect) bytes in 1 blocks are definitely lost in loss record 79 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x7920E33: Stream_New (in /usr/lib/x86_64-linux-gnu/libwinpr-utils.so.0.1.0)
==12578==    by 0x6A55E84: nego_recv_response (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56013: nego_attempt_tls (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56144: nego_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A71C42: rdp_client_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68A87: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4C9D5: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== 1,072 (48 direct, 1,024 indirect) bytes in 1 blocks are definitely lost in loss record 80 of 88
==12578==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12578==    by 0x7920E33: Stream_New (in /usr/lib/x86_64-linux-gnu/libwinpr-utils.so.0.1.0)
==12578==    by 0x6A55E84: nego_recv_response (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56073: nego_attempt_rdp (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A56144: nego_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A71C42: rdp_client_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x6A68A87: freerdp_connect (in /usr/lib/x86_64-linux-gnu/libfreerdp-core.so.1.1.0)
==12578==    by 0x4E4C9D5: xf_thread (in /usr/lib/x86_64-linux-gnu/libxfreerdp-client.so.1.1.0)
==12578==    by 0x8053181: start_thread (pthread_create.c:312)
==12578==    by 0x57F900C: clone (clone.S:111)
==12578== 
==12578== LEAK SUMMARY:
==12578==    definitely lost: 430 bytes in 10 blocks
==12578==    indirectly lost: 3,104 bytes in 4 blocks
==12578==      possibly lost: 0 bytes in 0 blocks
==12578==    still reachable: 45,583 bytes in 573 blocks
==12578==         suppressed: 0 bytes in 0 blocks
==12578== Reachable blocks (those to which a pointer was found) are not shown.
==12578== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==12578== 
==12578== For counts of detected and suppressed errors, rerun with: -v
==12578== ERROR SUMMARY: 16 errors from 11 contexts (suppressed: 0 from 0)

**Comment by bmiklautz** [comment]
@execlogic is the host you are connecting to a broker? If not the following looks strange to me:
```text
connected to 192.168.20.100:3389
connected to 192.168.20.100:3389
connected to 192.168.20.100:3389
```

**Comment by hardening** [comment]
@bmiklautz I think it's the nego workflow that first try in NLA, then TLS and fallback to RDP security

**Comment by bmiklautz** [comment]
@hardening stable-1.1. Interesting - I tried to reproduce this but no success (tried with all negos).
@execlogic  what windows version are you connecting to?

**Comment by execlogic** [comment]
@bmiklautz Windows Server 2003 R2 Standart x64 Edtition Service Pack 2.

**Comment by execlogic** [comment]
@bmiklautz This crash occurs while working with the accounting program. After 5-10 minutes in active work.

**Comment by execlogic** [comment]
@bmiklautz Can record video with the crash.

**Comment by bmiklautz** [comment]
@execlogic thank you for the information. I think there is no video required.

**Comment by bmiklautz** [comment]
@execlogic this is the same problem as reported in issue #1500 . This was already fixed on master.
If you build yourself I'd recommend to try out master. If you use a packaged version you might want to try a nightly build (https://ci.freerdp.com/job/freerdp-nightly-binaries/) to see if it is fixed.

**Comment by execlogic** [comment]
@bmiklautz 
This is commit, im right?
https://github.com/bmiklautz/FreeRDP/commit/c99d9ee72bae06d19a15cce46eb4f3230a97f296

**Comment by bmiklautz** [comment]
@execlogic exactly.
I've already created/merged PR #2440. So it's now fixed on stable-1.1 as well.

**Comment by execlogic** [comment]
@bmiklautz Thank you!!!
